### PR TITLE
init score utility with BLEU support

### PR DESCRIPTION
**Pull request by monsieurzhang**
local testing OK.

### CODE REVIEW

File: utilities/score/entrypoint.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import os
```
Comment (guillaumekln) [inline utilities/score/entrypoint.py]: Please remove all unused imports.

File: utilities/score/entrypoint.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import six
+import re
+import time
+import random
+import subprocess
+import json
+
+from nmtwizard.utility import Utility
```
Comment (guillaumekln) [inline utilities/score/entrypoint.py]: Same here.

File: utilities/score/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+source utilities/score/env/bin/activate
+pip install -r utilities/score/requirements.txt
+```
+
+2\. Define local environment:
+
+```bash
+export MODELS_DIR=/tmp/models
+export WORKSPACE_DIR=/tmp/workspace
+export CORPUS_DIR=$PWD/test/corpus
```
Comment (guillaumekln) [inline utilities/score/README.md]: Remove `MODELS_DIR` and `CORPUS_DIR` that are not relevant to this utility.

File: utilities/score/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,41 @@
+FROM nvidia/cuda:9.0-devel-ubuntu16.04 as cuda_devel
```
Comment (guillaumekln) [inline utilities/score/Dockerfile]: This looks like the Docker for similarity?

File: utilities/score/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+export TOOLS_DIR=$PWD/utilities/score
+export PYTHONPATH=$PWD:$PYTHONPATH
+```
+
+3\. Run:
+
+### Training
+
+```bash
+python utilities/score/entrypoint.py score -i ${CORPUS_DIR}/eval/tgt.txt -r ${CORPUS_DIR}/eval/ref.txt -l en
```
Comment (guillaumekln) [inline utilities/score/README.md]: Replace `${CORPUS_DIR}` by `test/corpus`.

File: utilities/score/entrypoint.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        self._tools_dir = os.getenv('TOOLS_DIR', '/root/tools')
+
+    @property
+    def name(self):
+        return "score"
+
+    def declare_arguments(self, parser):
+        subparsers = parser.add_subparsers(help='Run type', dest='cmd')
+        parser_score = subparsers.add_parser('score', help='Evaluate translation.')
+        parser_score.add_argument('-i', '--input', required=True, nargs='?',
```
Comment (guillaumekln) [inline utilities/score/entrypoint.py]: Why `nargs='?'`? I think it can be removed.

File: utilities/score/entrypoint.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                                  help='Input file.')
+        parser_score.add_argument('-r', '--ref', required=True, nargs='?',
+                                  help='Reference file.')
+        parser_score.add_argument('-l', '--lang', default='en', nargs='?',
+                                  help='Lang ID')
+
+    def convert_to_local_file(self, nextval):
+        inputs = nextval.split(',')
```
Comment (guillaumekln) [inline utilities/score/entrypoint.py]: I don't think we want to support multiple files here, and certainly not via a comma-separated list. And `eval_BLEU` does not look like it supports multiple files anyway.

I suggest to just assume there is a single file.

File: utilities/score/entrypoint.py (hunk 1 above)
Comment (monsieurzhang) [reply utilities/score/entrypoint.py]: done

File: utilities/score/README.md (hunk 3 above)
Comment (monsieurzhang) [reply utilities/score/README.md]: done

File: utilities/score/Dockerfile (hunk 4 above)
Comment (monsieurzhang) [reply utilities/score/Dockerfile]: updated

File: utilities/score/README.md (hunk 5 above)
Comment (monsieurzhang) [reply utilities/score/README.md]: done

File: utilities/score/entrypoint.py (hunk 6 above)
Comment (monsieurzhang) [reply utilities/score/entrypoint.py]: done

File: utilities/score/entrypoint.py (hunk 7 above)
Comment (monsieurzhang) [reply utilities/score/entrypoint.py]: updated

File: utilities/score/entrypoint.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        parser_score.add_argument('-l', '--lang', default='en',
+                                  help='Lang ID')
+
+    def eval_BLEU(self, tgtfile, reffile):
+        result = subprocess.check_output('perl ' + self._tools_dir + '/BLEU/multi-bleu-detok_cjk.perl ' + reffile + ' < ' + tgtfile, shell=True)
+        bleu = re.match("^BLEU\s=\s([\d\.]+),", result.decode('ascii'))
+        return bleu.group(1)
+
+    def exec_function(self, args):
+        val_tgt = args.input
```
[the remark this answers is not in the document]
Comment (monsieurzhang) [reply utilities/score/entrypoint.py]: yes, but I haven't tested this function.

**Comment by guillaumekln** [comment]
@Jean, is it OK for you? If so, please merge the PR.

**Comment by jsenellart** [comment]
few changes required:
* I propose first to rename `-i` (=`--input`) to `-o` (=`--output`) to  be more clear on the nature of these files
* I would like also to support evaluation of multiple files - just like translation. `-i file1.out file2.out -r file1.ref file2.ref` so that we can batch evaluate instead of creating 50 micro-tasks. For that, the output should be a dictionary `{"file1.out":{"BLEU":xxx,...},"file2.out":{"BLEU":yyy,...}`
* for multi-reference, I propose to use comma-separated: `-r file1.ref1,file1.ref2`. Theoretically comma could be used in file name, but practically it is a minor restriction

in any cases, each of these files reference could be remote file with storage prefix.

**Comment by jsenellart** [comment]
good to go! Thanks!

**Pull request merged**